File: src/univariate/chi.jl
immutable Chi <: ContinuousUnivariateDistribution
    df::Float64
    function Chi(df::Real)
        df > zero(df) || error("df must be positive")
        new(float64(df))
    end
end

@continuous_distr_support Chi 0.0 Inf

cdf(d::Chi, x::Real) = cdf(Chisq(d.df),x^2)
ccdf(d::Chi, x::Real) = ccdf(Chisq(d.df),x^2)
logcdf(d::Chi, x::Real) = logcdf(Chisq(d.df),x^2)
logccdf(d::Chi, x::Real) = logccdf(Chisq(d.df),x^2)

quantile(d::Chi,p::Real) = sqrt(quantile(Chisq(d.df),p))
cquantile(d::Chi,p::Real) = sqrt(cquantile(Chisq(d.df),p))
invlogcdf(d::Chi,p::Real) = sqrt(invlogcdf(Chisq(d.df),p))
invlogccdf(d::Chi,p::Real) = sqrt(invlogccdf(Chisq(d.df),p))

mean(d::Chi) = √2 * gamma((d.df + 1.0) / 2.0) / gamma(d.df / 2.0)

function mode(d::Chi)
    d.df >= 1.0 || error("Chi distribution has no mode when df < 1")
    sqrt(d.df - 1)
end

var(d::Chi) = d.df - mean(d)^2

function skewness(d::Chi)
    μ, σ = mean(d), std(d)
    (μ / σ^3) * (1.0 - 2.0 * σ^2)
end

function kurtosis(d::Chi)
    μ, σ, γ = mean(d), std(d), skewness(d)
    (2.0 / σ^2) * (1 - μ * σ * γ - σ^2)
end

function pdf(d::Chi, x::Real)
    k = d.df
    (2.0^(1.0 - k / 2.0) * x^(k - 1.0) * exp(-x^2 / 2.0)) / gamma(k / 2.0)
end

function entropy(d::Chi)
    k = d.df
    lgamma(k / 2.0) - log(sqrt(2.0)) -
        ((k - 1.0) / 2.0) * digamma(k / 2.0) + k / 2.0
end

function gradloglik(d::Chi, x::Float64)
  insupport(Chi, x) ? (d.df - 1.0) / x - x : 0.0
end

rand(d::Chi) = sqrt(rand(Chisq(d.df)))

File: test/ref/continuous/chi.R

Chi <- R6Class("Chi",
    inherit = ContinuousDistribution,
    public = list(
        names = c("nu"),
        nu = NA,
        initialize = function(nu) {
            self$nu <- nu
        },
        supp = function() { c(0, Inf) },
        properties = function() {
            k <- self$nu
            u <- sqrt(2) * gamma((k+1)/2) / gamma(k/2)
            v <- k - u^2
            m3 <- u / v^1.5 * (1 - 2 * v)
            m4 <- (2 / v) * (1 - u * sqrt(v) * m3 - v)
            list(dof = k,
                 mode = ifelse(k >= 1, sqrt(k - 1), NaN),
                 mean = u,
                 var = v,
                 skewness = m3,
                 kurtosis = m4,
                 entropy = lgamma(k/2) + (k - log(2) - (k-1)*psigamma(k/2, 0)) / 2
                )
        },
        pdf = function(x, log=FALSE) { chi::dchi(x, self$nu, log=log) },
        cdf = function(x) { chi::pchi(x, self$nu) },
        quan = function(v) { chi::qchi(v, self$nu) }
    )
)
